pkg/apis/pipeline/v1/result_validation.go

```diff
@@ -66,3 +65,47 @@ func validateObjectResult(tr TaskResult) (errs *apis.FieldError) {
 	}
 	return nil
 }
+
+// validateValue validates the value of the TaskResult.
+// It requires that enable-step-actions is true, the value is of type string
+// and format $(steps.<stepName>.results.<resultName>)
+func (tr TaskResult) validateValue(ctx context.Context) (errs *apis.FieldError) {
+	if tr.Value != nil {
+		if !config.FromContextOrDefaults(ctx).FeatureFlags.EnableStepActions {
+			return apis.ErrGeneric("feature flag %s should be set to true to fetch Results from Steps using StepActions.", config.EnableStepActions)
+		}
+		if tr.Value.Type != ParamTypeString {
+			return &apis.FieldError{
+				Message: fmt.Sprintf(
+					"Invalid Type. Wanted string but got: \"%v\"", tr.Value.Type),
+				Paths: []string{
+					fmt.Sprintf("%s.type", tr.Name),
+				},
+			}
+		}
+		if tr.Value.StringVal != "" {
+			_, _, err := ExtractStepResultName(tr.Value.StringVal)
+			if err != nil {
+				return &apis.FieldError{
+					Message: fmt.Sprintf("Could not extract step and result name from TaskResult. Expect value to look like $(steps.<stepName>.results.<resultName>) but got %v. Got error: %v", tr.Value.StringVal, err),
+					Paths: []string{
+						fmt.Sprintf("%s.value", tr.Name),
+					},
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// ExtractStepResultName extracts the step name and result name from a string matching
+// formtat $(steps.<stepName>.results.<resultName>).
+// If a match is not found, an error is retured.
+func ExtractStepResultName(value string) (string, string, error) {
+	re := regexp.MustCompile(`\$\(steps\.(.*?)\.results\.(.*?)\)`)
```

Comment:
if my input is `$(steps.foo.foo.results.Bar)`, this will return foo.foo and Bar, but I think we will return validation error from other places right(in following PRs)? Like the result name invalid, non-existent variable etc.

Reply:
I fixed this. I validate that the step name and result name are valid after extracting. 

In principle they should be in the first place otherwise the web hook will complain that we declared invalid step and result names while validating them.